### PR TITLE
search from table, comfortable reused table

### PR DESCRIPTION
**Pull request by levil664**

### CODE REVIEW

File: apps/schools/domains/common/components/table/forming.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { ColumnType } from 'antd/lib/table/interface'
+import React from 'react'
+import { getSearchText } from '@domains/common/utils/searchText'
+
+export interface RawColumnType<DataType> extends ColumnType<DataType> {
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/forming.tsx]: what is DataType?

File: apps/schools/domains/common/components/table/forming.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    return baseColumns.map((column) => ({
+        ...column,
+    }))
+}
+
+function escapeRegExp(text: string) {
+    return text.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&')
+}
+
+function HighlightText({ text, searchText }: any) {
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/forming.tsx]: any?

File: apps/schools/domains/common/components/table/forming.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    dataIndex: string
+}
+
+export function useGenerateFullColumns<DataType>(baseColumns: RawColumnType<DataType>[]): ColumnType<DataType>[] {
+    return baseColumns.map((column) => ({
+        ...column,
+    }))
+}
+
+function escapeRegExp(text: string) {
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/forming.tsx:18]: strange name
can you explain?

File: apps/schools/domains/common/components/table/forming.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+function HighlightText({ text, searchText }: any) {
+    const isMatch = text.includes(searchText)
+
+    if (isMatch) {
+        const parts = text.split(new RegExp(`(${escapeRegExp(searchText)})`, 'gi'))
+        return (
+            <span>
+                {parts.map((part: any, index: any) =>
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/forming.tsx]: any?

File: apps/schools/domains/common/components/table/forming.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                    part.toLowerCase() === searchText.toLowerCase() ? (
+                        <span key={index} style={{ backgroundColor: 'yellow' }}>
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/forming.tsx]: this style should be in colors mixin
can we make it with class?

File: apps/schools/domains/common/components/table/interfaces.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { TableProps } from 'antd'
+import React from 'react'
+
+export interface CustomTableProps extends TableProps<DataType> {
+    customType?: 'tableWithSearch' | 'tableWithoutSearch'
+    columnsTitles: string[]
+    columnsKeys: string[]
+    data: any
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/interfaces.ts]: any?

File: apps/schools/domains/common/components/table/interfaces.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { TableProps } from 'antd'
+import React from 'react'
+
+export interface CustomTableProps extends TableProps<DataType> {
+    customType?: 'tableWithSearch' | 'tableWithoutSearch'
+    columnsTitles: string[]
+    columnsKeys: string[]
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/interfaces.ts]: make one list with pairs

File: apps/schools/domains/common/components/table/interfaces.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    data: any
+    isLoading: boolean
+    searchFields: string[]
+    searchRequestText: string
+    setSearchRequestText: React.Dispatch<React.SetStateAction<string>>
+    mainRoute: string
+}
+
+export interface DataType {
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/interfaces.ts]: strange name of interface

File: apps/schools/domains/common/constants/Table.ts [hunk 9]
```diff
@@ -0,0 +1,2 @@
+export const searchColumns = ['phone', 'position', 'name']
```
Comment (mzavarka) [inline apps/schools/domains/common/constants/Table.ts]: it is strange
will we write there all names of columns? I thouth we can define it in special compnents

File: apps/schools/domains/common/utils/calculateAverageWidth.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+export function calculateAverageWidth(columns: any) {
```
Comment (mzavarka) [inline apps/schools/domains/common/utils/calculateAverageWidth.ts]: any?

File: apps/schools/domains/common/utils/searchText.ts [hunk 11]
```diff
@@ -0,0 +1,7 @@
+export function getSearchText(text: string) {
+    return text.split(':')[0]
+}
+
+export function createSearchTextForRequest(text: string, searchColumns: string[]) {
+    return `${text}:[${searchColumns.map((e: string) => e)}]`
```
Comment (mzavarka) [inline apps/schools/domains/common/utils/searchText.ts]: is there , in separator role?

File: apps/schools/domains/employee/components/employeeList/index.tsx [hunk 12]
```diff
@@ -1,111 +1,47 @@
...
-        width: '30%',
-    },
-].filter((item) => !item.hidden)
+import { searchColumns } from '@domains/common/constants/Table'
+import { Table } from '@domains/common/components/table'
+import { createSearchTextForRequest } from '@domains/common/utils/searchText'
 
 export function EmployeeList() {
-    const [dataSource, setDataSource] = useState<any[]>([])
+    const [searchRequestText, setSearchRequestText] = useState('')
```
Comment (mzavarka) [inline apps/schools/domains/employee/components/employeeList/index.tsx:15]: can we locate this logic to tableWithSearch component?

**Review by mzavarka — CHANGES_REQUESTED** [review]
Strange names
A lot of any typing

File: apps/schools/domains/common/components/table/forming.tsx (hunk 3 above)
Comment (levil664) [reply apps/schools/domains/common/components/table/forming.tsx:18]: ?
image

File: apps/schools/domains/common/utils/searchText.ts (hunk 11 above)
Comment (levil664) [reply apps/schools/domains/common/utils/searchText.ts]: The separator in this case will be a comma, by default the map method separates them with commas

File: apps/schools/domains/employee/components/employeeList/index.tsx (hunk 12 above)
Comment (levil664) [reply apps/schools/domains/employee/components/employeeList/index.tsx:15]: It is not clear why, too much logic should be added to the component in order to replace useState with a constant. Keep It Simple, Stupid.

File: apps/schools/domains/common/components/table/forming.tsx (hunk 3 above)
Comment (mzavarka) [reply apps/schools/domains/common/components/table/forming.tsx:18]: smells like fire

File: apps/schools/domains/common/components/table/styles/styles.module.scss [hunk 13]
```diff
@@ -23,3 +24,7 @@
     opacity: .8;
   }
 }
+
+.highlightText {
+  background-color: $color-yellow
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/styles/styles.module.scss]: name it $color-search-highlight-text

File: apps/schools/domains/common/components/table/interfaces.ts [hunk 14]
```diff
@@ -1,21 +1,27 @@
...
     mainRoute: string
 }
 
-export interface DataType {
+export interface RowType {
     id: string
     name: string
     position: string
     phone: string
 }
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/interfaces.ts]: why this type there?

it is local type of specific page

File: apps/schools/domains/employee/components/employeeList/index.tsx [hunk 15]
```diff
@@ -5,13 +5,13 @@ import styles from './styles/styles.module.scss'
...
 import { useOrganization } from '@domains/organization/providers/organizationProvider'
 import { Button } from '@domains/common/components/button'
-import { searchColumns } from '@domains/common/constants/Table'
 import { Table } from '@domains/common/components/table'
 import { createSearchTextForRequest } from '@domains/common/utils/searchText'
 
 export function EmployeeList() {
     const [searchRequestText, setSearchRequestText] = useState('')
     const { organizationId } = useOrganization()
+    const searchColumns = ['phone', 'position', 'name']
```
Comment (mzavarka) [inline apps/schools/domains/employee/components/employeeList/index.tsx]: locate it to local constants

File: apps/schools/domains/common/components/table/interfaces.ts [hunk 16]
```diff
@@ -1,21 +1,27 @@
...
+import { ReturnedData } from '@domains/common/redux/interfaces'
 
-export interface CustomTableProps extends TableProps<DataType> {
+export interface CustomTableProps extends TableProps<RowType> {
     customType?: 'tableWithSearch' | 'tableWithoutSearch'
-    columnsTitles: string[]
-    columnsKeys: string[]
-    data: any
+    columnsTitlesAndKeys: Array<string[]>
+    data?: ReturnedData<GetListEmployee[]>
```
Comment (mzavarka) [inline apps/schools/domains/common/components/table/interfaces.ts]: you should use generic except of GetListEmployee

**Review by mzavarka — CHANGES_REQUESTED** [review]

**Review by mzavarka — APPROVED** [review]

**Pull request merged**